lib/fetching/content-services/comments-content-service.js

```diff
@@ -0,0 +1,170 @@
+        } else if (res.statusCode === 401) {
+            jwtToken = null;
+            this._sendRequest(url, opts, callback);
+            return;
+        } else if (res.statusCode !== 200) {
+            this.emit('error', new Error.HTTP(res.statusCode));
+            return callback([]);
+        }
+
+        var comments = body;
+        try {
+            if (typeof comments === 'string') {
+                comments = JSON.parse(comments);
+            }
+            if (!Array.isArray(comments)) {
+                this.emit('error', new Error('Wrong data'));
+                return callback([]);
+            }
+        } catch (e) {
+            this.emit('error', new Error('Parse error: ' + e.message));
+            return callback([]);
+        }
+
+        var remaining = comments.length;
+        var reports = [];
+
+        if (comments.length === 0) {
+            return callback([]);
+        }
+
+        this.cache = {};
+
+        const { acquiredAt, timestamp } = comments[comments.length - 1];
+
+        // TODO rewrite with async/await to avoid callback hell
+        comments.forEach((doc) => {
+            this._parse(doc, (report) => {
+                reports.push(report);
+                if (--remaining === 0) {
+                    config.update(
+                        'comments:lastFetchDate',
+                        this._lastFetchDate.getTime(),
+                        (err) => {
+                            if (err) {
+                                logger.error(err);
+                            }
+                            config.update(
+                                'comments:lastAuthoredDate',
+                                this._lastAuthoredDate.getTime(),
+                                (err) => {
+                                    if (err) {
+                                        logger.error(err);
+                                    }
+                                    this._lastFetchDate = new Date(acquiredAt);
+                                    this._lastAuthoredDate = new Date(timestamp);
+                                    callback(reports);
+                                }
+                            );
+                        }
+                    );
+                }
+            });
+        });
+    });
+}
+
+CommentContentService.prototype._parse = function(doc, callback) {
+
+    var metadata = {}
+    if (typeof doc.media === 'object') { 
+        const ext = doc.media.ext;
+        const type = (doc.media.ext === 'mp4') ? 'video' : 'photo';
+        const mediaUrl = completeUrl(`static/${doc._id}.${ext}`);
+        metadata.mediaUrl = [{ type, mediaUrl }];
+    }
+    var content = doc.text || 'No Content';
+    var { post: postUrl, url } = doc;
+
+    const report = {
+        authoredAt: new Date(doc.timestamp),
+        fetchedAt: new Date(),
+        content,
+        url,
+        metadata,
+    }
+
+    const post = this.cache[postUrl];
```

Comment:
This memoization can be great for performance, but do you know if it's really necessary? Generally `findOne` should be very fast assuming reports are indexed on `url`. I worry this might be premature optimization that adds complexity without explicit benefit.

Reply:
I am hesitant to remove the cache, for the reason that I think it will reduce what might be 100 `Report.findOne` calls to 1-3 on average. To explain further:

Assuming the comment acquisition system runs at full capacity, then `_parse` will be called 100 times every ten seconds for each of the 100 comments fetched from the API. On average, these 100 comments will be associated with just 1-3 posts. Thus, caching post reports fetched from the database will reduce 100 separate and mostly redundant indexed `Report.findOne` calls to 1-3.

@cooperka knowing the context better, is `Report.findOne` fast enough when indexed by URL such that even a search by URL through the hundreds of thousands of reports is trivial to execute? If so, I can happily remove this optimization.

Reply:
Finding via index should generally be `O(1)` (extremely fast) in all cases. **Searching** (as in, arbitrary regex match) is slower but I don't think you're doing that here.

Anyway, I assume it's currently working so let's keep as is. Don't want to throw a wrench in things this close to the election!

Reply:
I'll at least at a TODO note for the future. Did not know that indexed searches are O(1)!